auth/src/main/java/com/redhat/lightblue/rest/auth/ldap/LdapFindUserByUidCommand.java

```diff
@@ -1,62 +1,122 @@
+     */
+    @Override
+    protected SearchResult getFallback() {
+        LOGGER.info("Error during the execution of the command. Falling back to the cache");
+        return new FallbackViaLDAPServerProblemCommand(cacheKey, getFailedExecutionException()).execute();
 
-            //make sure there is not another item available, there should be only 1 match
-            if (results.hasMoreElements()) {
-                LOGGER.error("Matched multiple users for the accountName: " + uid);
+    }
+
+
+    /*
+        Use the cache in case tje LDAP Server was not available and also to we have metrics around the fallback
+     */
+    private static class FallbackViaLDAPServerProblemCommand extends HystrixCommand<SearchResult> {
+        private static final Logger LOGGER = Logger.getLogger(FallbackViaLDAPServerProblemCommand.class);
+
+        private final LDAPCacheKey cacheKey;
+        private final Throwable failedExecutionException;
+
+        public FallbackViaLDAPServerProblemCommand(LDAPCacheKey cacheKey, Throwable failedExecutionException) {
+            super(HystrixCommand.Setter.withGroupKey(HystrixCommandGroupKey.Factory.asKey(GROUPKEY)).
+                    andCommandKey(HystrixCommandKey.Factory.asKey(GROUPKEY + ":" + FallbackViaLDAPServerProblemCommand.class.getSimpleName())));
+            this.cacheKey = cacheKey;
+            this.failedExecutionException = failedExecutionException;
+        }
+
+        @Override
+        protected SearchResult run() throws Exception {
+            //iff it was a problem with the server, try the following:
+            if(failedExecutionException instanceof NamingException) {
+                SearchResult searchResult = LDAPCache.getLDAPCacheSession().getIfPresent(cacheKey);
+                if (searchResult == null) {
+                   // retry to reach teh server and update the cache as it is not present, in case of error, fail
+                    searchResult = LDAPCache.getLDAPCacheSession().get(cacheKey);
+                }
+                // was able to use the cache or use the LDAP server on the second retry
+                return searchResult;
+
+            } else if(failedExecutionException instanceof Exception) {
+                // if it is an exception, propagate it until someone handle it properly
+                throw ((Exception)failedExecutionException);
+            } else {
+                // probably java.lang.Error or a custom Throwable exception. In this case we can't handle the problem
+                // and if we wrap it we can change the JVM to inconsistent state, so we have to exit with error status
+                // code
+                LOGGER.fatal("a Throwable instance (which is not instance of Exception) was thrown and it can't be handled.",failedExecutionException);
+                System.exit(-1);
```

Comment:
Yeah, that was the most uncomfortable but necessary thing, I tried to briefly explain why I did that in the comment, and again in summary as we are dealing with Throwable and we tried all the Exceptions classes and related possibility, that can be an Error which is something wrong with the JVM or another not recoverable issue. Anotehr problem with this that lead em to do that is that Hystrix throws in the fallback back Throwable and in the run method it only accepts Execptions, making us to deal with the Throwable (as we can not re throw it by java lang spec) and if we wrap with an Execption, again, we can be dealing with out of memory and ignoring it would lead the JVM to not predictable state , e.g.